packages/eslint-plugin-svelte/src/utils/svelte-context.ts

```diff
@@ -0,0 +1,182 @@
+import type { RuleContext } from '../types.js';
+import fs from 'fs';
+import path from 'path';
+import { getPackageJson } from './get-package-json.js';
+import { getFilename, getSourceCode } from './compat.js';
+
+const isRunOnBrowser = !fs.readFileSync;
+
+export type SvelteContext = {
+	svelteVersion: string;
+	fileType: '.svelte' | '.svelte.[js|ts]';
```

Comment:
Hmm, I don't like the design of having separate `fileType` and `svelteKitFileType`. I see 2 ways to do it better:
1. Merge them
2. Don't provide these, but go conceptually one level further - why would we want to know the file type? Maybe what we actually want to know is things like
    1. Are we in a SFC file?
    2. Are we in the server context?
    3. etc.

I think the second solution is more work (and requires quite careful design), but ultimately serves us better, because rules are fundamentally going to be enabled based on the *properties* of the current context/file, not based on its type...

Reply:
At here, it’s better to keep things simple and store only basic information. This is because the definition of "context" can change depending on the rules. For example, we might want to create a rule that prevents placing files like `xxx.svelte.js` in the same directory as `+page.svelte`.
Another example is that `xxx.svelte.js` might be used for implementing composables, and we might want to enforce a rule that their names must start with `useX`.

In short, since the definition of "context" depends on the rules, it’s better to just manage file types here.

Reply:
I'm still not convinced. I think we should instead add something like a `fileContext` field where we add some information about the file and we can possibly extend it in the future. I am not sure your first example is possible in eslint, as it operates on files separately. But anyway, we would probably need a separate flag for `.svelte.js/ts` files. But I don't think it is useful to have a way to distinguish `+page.js`, `+page.server.js`, `+layout.js` and `+layout.server.js` - what you would actually want in most situations is 2 flags - "Is this a layout file?" and "Is this a server-only file?"

Does that make sense to you or am I being too convoluted?

Reply:
This is one of specific example why "Is this a layout file?" and "Is this a server-only file" is not enough.
https://github.com/sveltejs/eslint-plugin-svelte/issues/438

This rule should check only `+layout.(js|ts)`. (`+layout.svelte` should not check.)

Reply:
Hmm, maybe I'm wrong (I haven't used SvelteKit server-side yet), but #438 should also check `+page.(js|ts)`, no? Then it's exactly what I've been getting at :D The differentiator isn't really a list of filename suffixes, but a role the file plays in SvelteKit...

Anyway, if my change is not accepted, then I would like to think a little bit more about how `fileType` and `svelteKitFileType` interact...

Reply:
And yes, the list of "file role flags" would probably be longer and would need to expand over time and some rules would check multiple flags at once...

Reply:
> And yes, the list of "file role flags" would probably be longer and would need to expand over time and some rules would check multiple flags at once...

Yes. In that case, we can add helper functions for common validation patterns.
I think it’s best to start with basic functionality and then add more convenient features later if they prove to be valuable.

Reply:
> Hmm, maybe I'm wrong (I haven't used SvelteKit server-side yet), but https://github.com/sveltejs/eslint-plugin-svelte/issues/438 should also check +page.(js|ts),

Yes. We need to check `+page.(js|ts)` also, but what I wanted to say is that "Is this a layout file?" and "Is this a server-only file" is not enough.

Reply:
> Yes. In that case, we can add helper functions for common validation patterns.

Hmm, yeah, that's probably a good enough alternative.

I was thinking about the `fileType` a little bit more as well. I get that in `svelteKitFileType`, the list of file suffixes makes sense, but I think it gets confusing with `fileType` pretty quickly - if I understand it correctly, a `+page.svelte` would havea file type `.svelte` and `+page.js` would have a file type of `.svelte.js` - which is a little counterintuitive. Maybe `fileType` could have values `"component"|"js"` or something similar? Then it wouldn't give the illusion of being a filename extension, which it isn't in case of SvelteKit files...